### DISCUSSION TITLE
Make `ValueEditorProps` use a generic Field type

### DISCUSSION THREAD

**Discussion by sneakytowelsuit** [react-querybuilder/react-querybuilder#406] [Ideas]
# General Idea
I use this component at work to support building a react hosted UI that allows users to manage Drools rules and have built my own translation layer from a combination of the provided `formatQuery` and `transformQuery` options that serves all my needs. 

What I have had to customize, and thankfully haven't had much of an issue with, is customizing some of the types to have stronger typed custom field names and operators, which are built from a combination of TypeScript enums with string values and template string literal types. This works great since in we need to enforce that field names match a set of known values, and I've managed this with a custom type that looks like
```typescript
enum AllowedClasses {
  SOME_CLASS = 'SomeClass',
  SOME_OTHER_CLASS = 'SomeOtherClass'
}

enum SomeClassFields {
  NAME = 'name',
  AGE = 'age'
}

enum SomeOtherClassFields {
  TASTE = 'taste',
  SMELL = 'smell'
}

/*
Create string literal template type from every permutation of C and F
*/
type AllowedFieldNames<C extends AllowedClasses, F extends SomeClassFields | SomeOtherClassFields> = `${C}.${F}`

type SomeClassFieldNames = AllowedFieldNames<AllowedClasses.SOME_CLASS, SomeClassFields>;
//         ^? type SomeClassFieldNames = 'SomeClass.name' | 'SomeClass.age'

type SomeOtherClassFieldNames = AllowedFieldNames<AllowedClasses.SOME_OTHER_CLASS, SomeOtherClassFields>;
//         ^? type SomeOtherClassFieldNames = 'SomeOtherClass.taste' | 'SomeOtherClass.smell'

type CustomField = Field & {
  name: SomeClassFieldNames | SomeOtherClassFieldNames
}
```
And by using this `CustomField` type we can enforce that field names match an expected format using the mapped type. I do a similar thing for the operators, since using custom operators makes translating to the Drools language format significantly easier. 

Since the `Field` type provided by the library extends `NameLabelPair` and as such allows for a field of type `[x: string]: any`, this means we can add new fields to our `CustomField` type that we can leverage later on. 

The first use case I can think of for this is when customizing the ValueEditor, where we can build custom editing experiences for the users based on potentially a combination of the field name and perhaps a `dataType` field [similar to the example in the documentation](https://react-querybuilder.js.org/docs/next/tips/custom-with-fallback).

# Proposed solution
Change the `ValueEditorProps` interface to
```typescript
export interface ValueEditorProps<F extends Field = Field> extends SelectorOrEditorProps {
  field: F['name'];
  operator: F['operators'] extends Array<Exclude<F['operators'], undefined>> ? F['operators'][number] : undefined;
  value?: any;
  valueSource: ValueSource;
  fieldData: F;
  type?: ValueEditorType;
  inputType?: string | null;
  values?: any[];
  listsAsArrays?: boolean;
}
```
Which yields the following benefits:
- A customized `Field` type similar to what we are using would be able to pass the template strings literals directly into the ValueEditorProps 
- The operators (which are also customized in our case) would be typed against the same custom field type
- The `fieldData` field would then have access to the same values

In the [example demonstrating how to use custom components]([similar to the example in the documentation](https://react-querybuilder.js.org/docs/next/tips/custom-with-fallback), this would give IntelliSense for the developer if the had a type that looked like
```typescript
type CustomField = Field & { datatype: 'date' | 'dateRange' | null }

const CustomValueEditor: FC<ValueEditorProps<CustomField>> = (props) => {
  if (props.fieldData.datatype === 'date') {
//                                      ^? 'date' | 'dateRange' | null
    // return if date here
  } 
  if(props.fieldData.datatype === 'dateRange') {
    // return if dateRange here
  }
  //default return
}
```

Pretty much all the other types from this library support generics well enough for me to pass around the customized types I've built, this is the last use case I can think of currently. I've already forked the repo and can open a PR with the changes if you'd like, unless there's another edge case I'm missing that this would break.

**Comment by jakeboone02** [comment]
Hi @Austin-Stowe! Thanks for all the detail here, it's really helpful. I'm especially glad you find `formatQuery` and `transformQuery` useful. A lot of thought and work went into those.

I think you have the beginning of a good idea, but I'd like to flesh it out a little. We should probably add generics for more than just the field-related props in `ValueEditorProps`. I'm thinking about `operator`, `value`, and `values` for starters. And I'm sure there are plenty of other places where generics would be useful. `OperatorSelectorProps` comes to mind first since it also has field-related props.

One note about your `operator` typings: the `Field#operators` property isn't the only place to define the possible operators for a field. You also have to take into account the `getOperators` function prop on the `<QueryBuilder />` itself, so the `F['operators']`-based code you've got there is probably too limited and/or limiting.

Why don't you go ahead and submit the PR and we can work on it from there. (Undo the modifications to `.all-contributorsrc` first; I run a script to update that and the README after merging PRs from new contributors.)

**Reply by sneakytowelsuit** [reply]
Absolutely, the `formatQuery` and `transformQuery` functions are absolutely essential to the workflow I'm creating and I can't express enough how grateful I am those exist. 

I'll go ahead and start a new branch using this discussion number and open a PR with some quick changes. I'll also undo the changes to the contributors section, I was mostly playing around in there to see how that worked.

**Reply by sneakytowelsuit** [reply]
[Draft PR for this feature](https://github.com/react-querybuilder/react-querybuilder/pull/408)

We can make whatever other changes this incurs on that PR, while the changes I've made would be enough to support my particular use case if we can make the generics more flexible to account for things like the `getOperators` function prop then I'm game. I'll start digging a bit deeper into the code to see if anything else stands out to me

**Reply by jakeboone02** [reply]
Thanks so much! I'll take a closer look in the next few days.

**Comment by jakeboone02** [comment]
I was going through old issues and discussions looking for something else, and I came across this one and realized it was the initial driving factor for the huge [TypeScript improvements](https://react-querybuilder.js.org/docs/migrate#typescript-updates) in RQB v7. Thanks for kicking off the process, @Austin-Stowe!

**Reply by sneakytowelsuit** [reply]
Happy to be a part of this library! It's been a huge help, and reading through those migration notes I'm impressed at the overhaul as well! Super neat to see how all the different types are able to infer from one another now, just means we get an even stronger DX than the already great level we had before.
